distribution/lib/Standard/Base/0.0.0-dev/src/Error/Extensions.enso

```diff
@@ -156,3 +173,121 @@ Error.is_error = True
          example_rethrow = Panic.rethrow Examples.throw_error
 Panic.rethrow : (Any ! Any) -> Any
 Panic.rethrow value = value.catch Panic.throw
+
+## Returns the stack trace of the caught panic.
+Caught_Panic.stack_trace : Vector.Vector Stack_Trace_Element
+Caught_Panic.stack_trace =
+    Panic.get_attached_stack_trace this
+
+## UNSTABLE
+   ADVANCED
+
+   Catches panics caused by a Java exception which is a subclass of the provided
+   class.
+
+   Arguments:
+   - action: the action to attempt and catch exceptions from.
+   - java_class: the class of a Java exception that should be caught.
+   - handler: a callback which is called with the caught exception.
+
+   Returns the result of running `action` or the result of the `handler` if the
+   matching Java exception has been thrown from within the `action`.
+
+   > Example
+     Convert a string to an integer, catching `NumberFormatException` and
+     converting it to a more Enso-friendly dataflow error.
+
+     polyglot java import java.lang.Long
+     polyglot java import java.lang.NumberFormatException
+     parse str =
+         Panic.catch_java_exception NumberFormatException (Long.parseLong str) exception->
+             Error.throw (Illegal_Argument_Error "The provided string is not a valid number: "+exception.getMessage)
+Panic.catch_java_exception : Any -> Any -> (Any -> Any) -> Any
```

Comment:
This is ugly, why is it not handled through `catch`?

Reply:
Because the direct type of a Java exception caught in Enso is `Polyglot_Error`. We cannot specify a type there, also we have a different kind of check.

Or do you suggest we should somehow check that the type is a Java class and then execute the logic from this function? Can we even check if a type is a Java class?